packages/edit-site/src/components/style-book/index.js

```diff
@@ -196,15 +205,43 @@ function StyleBook( {
 		[ examples ]
 	);
 
+	const { baseConfig } = useSelect(
+		( select ) => ( {
+			baseConfig:
+				select(
+					coreStore
+				).__experimentalGetCurrentThemeBaseGlobalStyles(),
+		} ),
+		[]
+	);
+
+	const mergedConfig = useMemo( () => {
+		if ( ! isObjectEmpty( userConfig ) && ! isObjectEmpty( baseConfig ) ) {
+			return mergeBaseAndUserConfigs( baseConfig, userConfig );
+		}
+		return {};
+	}, [ baseConfig, userConfig ] );
+
+
+	// Copied from packages/edit-site/src/components/revisions/index.js
+	// could we create a shared hook?
 	const originalSettings = useSelect(
 		( select ) => select( blockEditorStore ).getSettings(),
 		[]
 	);
+
 	const settings = useMemo(
 		() => ( { ...originalSettings, __unstableIsPreviewMode: true } ),
 		[ originalSettings ]
 	);
 
+	const [ globalStyles ] = useGlobalStylesOutputWithConfig( mergedConfig );
```

Comment:
This repeats what `<Revisions />` does. Should perhaps abstract into a hook

Reply:
Okay, the refactor goes deep 😄 

I might do it in a follow up. 

The plan:

-  add and export a new function `useGlobalStylesOutputWithCustomConfig( customConfig: GlobaStylesObject )` or something to [use-global-styles-output.js](https://github.com/WordPress/gutenberg/blob/5dec32e31ceaf8d87f527a348533ad7a0c7f5f8c/packages/block-editor/src/components/global-styles/use-global-styles-output.js)
- Move and export the util `mergeBaseAndUserConfigs` from [global-styles-provider.js](https://github.com/WordPress/gutenberg/blob/badc3441f1bd94fdad93825a188694b093ebf403/packages/edit-site/src/components/global-styles/global-styles-provider.js) to use-global-styles-output.js and update all usages.
- Use `useGlobalStylesOutputWithCustomConfig` in Revisions and Style Book

Note to self - working diff:

<details>

<summary>Diff</summary>

```diff
diff --git a/packages/block-editor/src/components/global-styles/index.js b/packages/block-editor/src/components/global-styles/index.js
index 76a95357ba..e7f965fe24 100644
--- a/packages/block-editor/src/components/global-styles/index.js
+++ b/packages/block-editor/src/components/global-styles/index.js
@@ -9,6 +9,7 @@ export {
 	getLayoutStyles,
 	useGlobalStylesOutput,
 	useGlobalStylesOutputWithConfig,
+	useGlobalStylesOutputWithCustomConfig,
 } from './use-global-styles-output';
 export { GlobalStylesContext } from './context';
 export {
diff --git a/packages/block-editor/src/components/global-styles/use-global-styles-output.js b/packages/block-editor/src/components/global-styles/use-global-styles-output.js
index 7e99eca355..37996587de 100644
--- a/packages/block-editor/src/components/global-styles/use-global-styles-output.js
+++ b/packages/block-editor/src/components/global-styles/use-global-styles-output.js
@@ -1,3 +1,8 @@
+/**
+ * External dependencies
+ */
+import deepmerge from 'deepmerge';
+
 /**
  * WordPress dependencies
  */
@@ -38,6 +43,7 @@ import {
 	setImmutably,
 } from '../../utils/object';
 import BlockContext from '../block-context';
+import { isPlainObject } from 'is-plain-object';
 
 // List of block support features that can have their related styles
 // generated under their own feature level selector rather than the block's.
@@ -48,6 +54,23 @@ const BLOCK_SUPPORT_FEATURE_LEVEL_SELECTORS = {
 	typography: 'typography',
 };
 
+
+function isObjectEmpty( object ) {
+	return ! object || Object.keys( object ).length === 0;
+}
+
+// Copied from packages/edit-site/src/components/global-styles/global-styles-provider.js
+// @todo: this should be exported and used from here?
+export function mergeBaseAndUserConfigs( base, user ) {
+	return deepmerge( base, user, {
+		// We only pass as arrays the presets,
+		// in which case we want the new array of values
+		// to override the old array (no merging).
+		isMergeableObject: isPlainObject,
+	} );
+}
+
+
 function compileStyleValue( uncompiledValue ) {
 	const VARIABLE_REFERENCE_PREFIX = 'var:';
 	const VARIABLE_PATH_SEPARATOR_TOKEN_ATTRIBUTE = '|';
@@ -1264,3 +1287,35 @@ export function useGlobalStylesOutput() {
 	const { merged: mergedConfig } = useContext( GlobalStylesContext );
 	return useGlobalStylesOutputWithConfig( mergedConfig );
 }
+
+/**
+ * Like useGlobalStylesOutputWithConfig but returns global style output by
+ * merging incoming "overrides" with base config and settings.
+ *
+ * The use case is to output real-time changes to global styles based on custom configs
+ * other than `userConfig`, for example historical global styles changes (revisions).
+ *
+ * @param {Object} customConfig A custom global styles configuration.
+ * @return {Array|undefined}    Array of global styles stylesheets and settings. Undefined if no styles are output.
+ */
+export function useGlobalStylesOutputWithCustomConfig( customConfig ) {
+	const { base: baseConfig } = useContext( GlobalStylesContext );
+
+	const mergedConfig = useMemo( () => {
+		if (
+			! isObjectEmpty( customConfig ) &&
+			! isObjectEmpty( baseConfig )
+		) {
+			return mergeBaseAndUserConfigs( baseConfig, customConfig );
+		}
+		return {};
+	}, [ baseConfig, customConfig ] );
+
+	const [ globalStyles ] = useGlobalStylesOutputWithConfig( mergedConfig );
+
+	if ( ! isObjectEmpty( globalStyles ) && ! isObjectEmpty( customConfig ) ) {
+		return globalStyles;
+	}
+
+	return undefined;
+}
diff --git a/packages/edit-site/src/components/revisions/index.js b/packages/edit-site/src/components/revisions/index.js
index aeb7779d06..ba1b53730d 100644
--- a/packages/edit-site/src/components/revisions/index.js
+++ b/packages/edit-site/src/components/revisions/index.js
@@ -6,52 +6,30 @@ import { __ } from '@wordpress/i18n';
 import {
 	BlockList,
 	privateApis as blockEditorPrivateApis,
-	store as blockEditorStore,
 	__unstableEditorStyles as EditorStyles,
+	store as blockEditorStore,
 	__unstableIframe as Iframe,
 } from '@wordpress/block-editor';
 import { useSelect } from '@wordpress/data';
 import { useMemo } from '@wordpress/element';
-import { store as coreStore } from '@wordpress/core-data';
 
 /**
  * Internal dependencies
  */
 
 import { unlock } from '../../lock-unlock';
-import { mergeBaseAndUserConfigs } from '../global-styles/global-styles-provider';
 import EditorCanvasContainer from '../editor-canvas-container';
 
-const { ExperimentalBlockEditorProvider, useGlobalStylesOutputWithConfig } =
-	unlock( blockEditorPrivateApis );
-
-function isObjectEmpty( object ) {
-	return ! object || Object.keys( object ).length === 0;
-}
+const {
+	ExperimentalBlockEditorProvider,
+	useGlobalStylesOutputWithCustomConfig,
+} = unlock( blockEditorPrivateApis );
 
 function Revisions( { actions, userConfig, blocks } ) {
-	const { baseConfig } = useSelect(
-		( select ) => ( {
-			baseConfig:
-				select(
-					coreStore
-				).__experimentalGetCurrentThemeBaseGlobalStyles(),
-		} ),
-		[]
-	);
-
-	const mergedConfig = useMemo( () => {
-		if ( ! isObjectEmpty( userConfig ) && ! isObjectEmpty( baseConfig ) ) {
-			return mergeBaseAndUserConfigs( baseConfig, userConfig );
-		}
-		return {};
-	}, [ baseConfig, userConfig ] );
-
 	const renderedBlocksArray = useMemo(
 		() => ( Array.isArray( blocks ) ? blocks : [ blocks ] ),
 		[ blocks ]
 	);
-
 	const originalSettings = useSelect(
 		( select ) => select( blockEditorStore ).getSettings(),
 		[]
@@ -60,13 +38,8 @@ function Revisions( { actions, userConfig, blocks } ) {
 		() => ( { ...originalSettings, __unstableIsPreviewMode: true } ),
 		[ originalSettings ]
 	);
-
-	const [ globalStyles ] = useGlobalStylesOutputWithConfig( mergedConfig );
-
 	const editorStyles =
-		! isObjectEmpty( globalStyles ) && ! isObjectEmpty( userConfig )
-			? globalStyles
-			: settings.styles;
+		useGlobalStylesOutputWithCustomConfig( userConfig ) || settings.styles;
 
 	return (
 		<EditorCanvasContainer
diff --git a/packages/edit-site/src/components/style-book/index.js b/packages/edit-site/src/components/style-book/index.js
index 2194920b18..4efd26a1f0 100644
--- a/packages/edit-site/src/components/style-book/index.js
+++ b/packages/edit-site/src/components/style-book/index.js
@@ -29,19 +29,17 @@ import { useSelect } from '@wordpress/data';
 import { useResizeObserver } from '@wordpress/compose';
 import { useMemo, useState, memo } from '@wordpress/element';
 import { ENTER, SPACE } from '@wordpress/keycodes';
-import { store as coreStore } from '@wordpress/core-data';
 
 /**
  * Internal dependencies
  */
 import { unlock } from '../../lock-unlock';
 import EditorCanvasContainer from '../editor-canvas-container';
-import { mergeBaseAndUserConfigs } from '../global-styles/global-styles-provider';
 
 const {
 	ExperimentalBlockEditorProvider,
 	useGlobalStyle,
-	useGlobalStylesOutputWithConfig,
+	useGlobalStylesOutputWithCustomConfig,
 } = unlock( blockEditorPrivateApis );
 
 const {
@@ -122,10 +120,6 @@ const STYLE_BOOK_IFRAME_STYLES = `
 	}
 `;
 
-function isObjectEmpty( object ) {
-	return ! object || Object.keys( object ).length === 0;
-}
-
 function getExamples() {
 	// Use our own example for the Heading block so that we can show multiple
 	// heading levels.
@@ -206,25 +200,6 @@ function StyleBook( {
 		[ examples ]
 	);
 
-	const { baseConfig } = useSelect(
-		( select ) => ( {
-			baseConfig:
-				select(
-					coreStore
-				).__experimentalGetCurrentThemeBaseGlobalStyles(),
-		} ),
-		[]
-	);
-
-	const mergedConfig = useMemo( () => {
-		if ( ! isObjectEmpty( userConfig ) && ! isObjectEmpty( baseConfig ) ) {
-			return mergeBaseAndUserConfigs( baseConfig, userConfig );
-		}
-		return {};
-	}, [ baseConfig, userConfig ] );
-
-	// Copied from packages/edit-site/src/components/revisions/index.js
-	// could we create a shared hook?
 	const originalSettings = useSelect(
 		( select ) => select( blockEditorStore ).getSettings(),
 		[]
@@ -235,12 +210,8 @@ function StyleBook( {
 		[ originalSettings ]
 	);
 
-	const [ globalStyles ] = useGlobalStylesOutputWithConfig( mergedConfig );
-
 	settings.styles =
-		! isObjectEmpty( globalStyles ) && ! isObjectEmpty( userConfig )
-			? globalStyles
-			: settings.styles;
+		useGlobalStylesOutputWithCustomConfig( userConfig ) || settings.styles;
 
 	return (
 		<EditorCanvasContainer


```

</details>